File: l5.py

s = 'babad'


def ispda(s):
    if s[::-1]== s:
        return True
    else:
        return False

def maxpda(s):
    outc = s[0]
    for i in range(len(s)):
        if s[i] not in s[i:]:
            continue
        else:
            j = len(s[i:])
            while j > 0:
                if s[i + j - 1] == s[i]:
                    iterm = s[i:i + j]
                    if iterm[::-1] == iterm:
                        if len(outc)<len(iterm):
                           outc = iterm
                        break
                j -= 1

    return outc

print(maxpda(s))


# an very interesting solution of this problem
class Solution:
    # @return a string
    def longestPalindrome(self, s):
        if len(s)==0:
            return 0
        maxLen=1
        start=0
        for i in xrange(len(s)):
            if i-maxLen >=1 and s[i-maxLen-1:i+1]==s[i-maxLen-1:i+1][::-1]:
                start=i-maxLen-1
                maxLen+=2
                continue

            if i-maxLen >=0 and s[i-maxLen:i+1]==s[i-maxLen:i+1][::-1]:
                start=i-maxLen
                maxLen+=1
        return s[start:start+maxLen]

File: l5.java
class Solution {
    public String longestPalindrome(String s) {
        int[][] dp = new int[s.length()][s.length()];
        int maxstart = 0;
        int maxend = 0;

        dp[0][0] = 1;
        for (int i=1; i < s.length(); i++) {
            dp[i][i] = 1;
            if (s.charAt(i) == s.charAt(i-1)) {
                dp[i-1][i] = 2;
                maxstart = i-1;
                maxend = i;
            }
        }


        for (int i=s.length()-1; i <= 0; i--) {
            for (int j=i+2; j < s.length(); j++) {
                if (s.charAt(i) == s.charAt(j) && dp[i+1][j-1] > 0) {
                    dp[i][j] = dp[i+1][j-1] + 2;
                    if (dp[i][j] > maxend - maxstart + 1) {
                        maxend = j;
                        maxstart = i;
                    }
                }
            }
        }

        return s.substring(maxstart, maxend + 1);
    }
}